File: frontend/src/services/billing.service.ts
import { apiClient } from '@/api/axios';
import { API_ENDPOINTS } from '@/constants/apiEndpoints';
import type {
  ApiResponse,
  CreateOrderRequest,
  PaymentOrder,
  Plan,
  Quota,
  Subscription,
  VerifyPaymentRequest,
} from '@/types';

/**
 * Billing service (billing-service microservice).
 *
 * Payment safety: quota is only upgraded after the server verifies the
 * Razorpay signature — a frontend callback alone never grants storage.
 */
export const billingService = {
  getPlans: () => apiClient.get<ApiResponse<Plan[]>>(API_ENDPOINTS.billing.plans),

  getSubscription: () =>
    apiClient.get<ApiResponse<Subscription>>(API_ENDPOINTS.billing.subscription),

  getQuota: () => apiClient.get<ApiResponse<Quota>>(API_ENDPOINTS.billing.quota),

  getOrderHistory: () =>
    apiClient.get<ApiResponse<PaymentOrder[]>>(API_ENDPOINTS.billing.orders),

  /** Creates a payment order (503 when the payment provider is not configured). */
  createOrder: (body: CreateOrderRequest) =>
    apiClient.post<ApiResponse<PaymentOrder>>(API_ENDPOINTS.billing.createOrder, body, {
      silent: true,
      // A 401 here is a gateway/session edge case, never the payment result —
      // keep the payment UI in control of its own error state.
      skipAuthRedirect: true,
    }),

  /** Confirms a payment after checkout — verified server-side by signature. */
  verifyPayment: (body: VerifyPaymentRequest) =>
    apiClient.post<ApiResponse<PaymentOrder>>(API_ENDPOINTS.billing.verifyOrder, body, {
      silent: true,
      skipAuthRedirect: true,
    }),

  /** Cancels a pending order (user abandoned the checkout). */
  cancelOrder: (orderUuid: string) =>
    apiClient.post<ApiResponse<PaymentOrder>>(API_ENDPOINTS.billing.cancelOrder(orderUuid)),
};

File: backend/billing-service/src/main/java/com/cloudnest/billing/service/BillingService.java
package com.cloudnest.billing.service;

import com.cloudnest.billing.dto.PaymentOrderResponse;
import com.cloudnest.billing.dto.PlanResponse;
import com.cloudnest.billing.dto.QuotaResponse;
import com.cloudnest.billing.dto.SubscriptionResponse;
import com.cloudnest.billing.dto.VerifyPaymentRequest;
import com.cloudnest.billing.entity.PlanType;

import java.util.List;

/**
 * Core billing operations: plans, subscriptions, payment orders,
 * payment verification, webhook handling and quota resolution.
 */
public interface BillingService {

    List<PlanResponse> getActivePlans();

    SubscriptionResponse getCurrentSubscription(Long userId);

    /** Creates a payment order for a plan upgrade (503 when provider unconfigured). */
    PaymentOrderResponse createOrder(Long userId, PlanType planType);

    /**
     * Verifies a provider payment confirmation and — only when the
     * signature validates — marks the order PAID and upgrades the plan.
     */
    PaymentOrderResponse verifyPayment(Long userId, VerifyPaymentRequest request);

    PaymentOrderResponse cancelOrder(Long userId, String orderUuid);

    List<PaymentOrderResponse> getOrderHistory(Long userId);

    /** Storage quota granted by the user's current subscription. */
    QuotaResponse getQuota(Long userId);

    /**
     * Handles a Razorpay webhook event after verifying its signature.
     * Idempotent: repeated events never double-credit quota.
     */
    void handleRazorpayWebhook(String rawBody, String signature);
}
